docs/cip/CIP-8_Matrix_Packages.md

```diff
@@ -0,0 +1,209 @@
+
+Ideally, the matrix package format could eventually be shared by the many tools in the emergent LLM/AI stack. This proposal seeks to balance that Chroma-agnostic goal with the reality that working, useful software is usually prerequisite to bootstrapping an ecosystem.
+## Impact
+
+This is a novel user path for Chroma, and as such does not affect any existing code or user behaviors. It may eventually affect workflows for developing and deploying Chroma-based apps, possibly even those that have no external dependencies.
+
+The main impact is on maintenance burden and (possible) compatibility guarantees. By creating what can be thought of as a portable serialization format for most of what is ordinarily stored in Chroma, it is likely that some properties which previously could have been considered implementation details, now will be assumed by users to be part of Chroma's public API surface.
+
+TODO examples
+
+## Proposed Change
+
+This is a proposal in two parts. First, we describe a package format. Then, we outline new chroma client capabilities and corresponding CLI subcommands that can round trip to and from this package format.
+
+Note that this CIP explicitly omits any actual dependency management behavior. While certain design choices here are motivated by preparing for that problem, directly addressing it is a problem for a future CIP.
+### Matrix Package Format
+A matrix package is a filesystem format consisting of two files, co-occurring within the same directory. Each file serves a particular, well-defined role:
+
+- `matrix.package.yaml`: defines the package name, version, and other common metadata, and certain metadata about the dataset in the package. Future iterations will add dependency declarations - the same kind of information that appears in `package.json` or `requirements.txt`. **Required**.
+- `matrix.data.parquet`: contains the package's dataset. Each row closely mirrors Chroma db content, containing original document source, computed embeddings, and metadata. **Required**, but will likely become optional in the future (more below)
+
+TODO - how do we serialize a computed index? it doesn't have the same cardinality as the data, does it? so that'd mean a separate file
+
+Most critical to note is that a matrix package contains *exactly one* dataset (see further discussion under [alternatives](#Multi-dataset%20packages.md)). For now, we can assume a 1:1 relationship between a matrix package and a Chroma collection. This simple correspondence may change once we allow dependencies between matrix packages.
```

Comment:
How good is a single Collection from a developer experience perspective? Are there perhaps some metrics in Posthog to inform such a decision @HammadB? When this CIP becomes a reality people WILL use it for backup and restore and having to export 100 or 1000 of collections into packages might not be great experience, we're just creating a problem looking for a solution :)

Reply:
I agree that you could abuse this for backup/restore but my off the cuff reaction is to treat them as separate things.

Reply:
Yes yes yes - the overlap here with backup/restore is really interesting. i do think it's a distinct case (as i'm trying to lead towards with [this question](https://github.com/chroma-core/chroma/pull/1179#discussion_r1337023118) about ids), but that likely won't stop folks from trying.

This comment thread makes me feel that perhaps the biggest omission from this initial pass at this doc is at least some discussion of possible relations between collection and matrix package/dataset - 1:1, 1:N, N:1, N:N, N:M - and when/why each might be appropriate.